returnn/tf/util/data.py

```diff
@@ -1164,26 +1185,15 @@ def maybe_unbroadcast_placeholder():
       """
       if res.placeholder is None:
         return
+      assert dim_tag.kind != DimensionTag.Types.Batch
       assert res.batch_shape[axis] == 1
       with tf.name_scope("copy_add_dim_by_tag_unbroadcast"):
-        dyn_size = None
-        if dim_tag.src_data is not None and dim_tag.src_axis is not None and dim_tag.src_data.placeholder is not None:
-          dyn_size = tf.shape(dim_tag.src_data.placeholder)[dim_tag.src_axis]
-        if dyn_size is None:
-          dyn_size = dim_tag.dimension
-        if dyn_size is None and dim_tag.dyn_size is not None:
-          dyn_size = tf.math.reduce_max(dim_tag.dyn_size)
-        if dyn_size is None and dim_tag.kind == DimensionTag.Types.Batch:
-          from returnn.tf.layers.base import LayerBase
-          dyn_size = LayerBase.get_recent_layer().get_batch_dim()
-        assert dyn_size is not None, 'need placeholder, dim_tag.dimension or dim_tag.dyn_size for unbroadcast'
-        tiles = [1] * axis + [dyn_size] + [1] * (res.batch_ndim - axis - 1)
+        tiles = [1] * axis + [dim_tag.get_unbroadcast_dim()] + [1] * (res.batch_ndim - axis - 1)
         res.placeholder = tf.tile(res.placeholder, tiles)
 
     if dim_tag.kind == DimensionTag.Types.Batch:
-      res = self.copy_add_batch_dim(batch_dim_axis=axis)
-      if unbroadcast:
-        maybe_unbroadcast_placeholder()
+      # Always unbroadcast batch dim axis
+      res = self.copy_add_batch_dim(batch_dim_axis=axis, batch_dim=dim_tag.get_unbroadcast_dim())
```

Comment:
Hm, now I see. I'm actually reconsidering whether we should do that here.
Maybe better:
```suggestion
      res = self.copy_add_batch_dim(batch_dim_axis=axis, batch_dim=dim_tag.get_dim_value() if unbroadcast else 1)
```
This is for efficiency. E.g. when you do `a *  b` (in terms of layers, e.g. via `CombineLayer`), and `b` doesn't have a batch-dim, and then this adds the dimensions (with `unbroadcast=False`), it would be way more efficient to not unbroadcast the batch-dim (both in terms of computation (unnecessary copying) and in terms of memory).

Reply:
This is somewhat inconsistent to what I said before. But this should really only play a role for an intermediate `Data` instance. A layer output would never have that. So I don't think this is a problem here.

Reply:
(This should also fix the failing test cases.)

Reply:
Okay sure! Let's do that then.